unix/vncconfig/vncconfig.man
.TH vncconfig 1 "" "TigerVNC" "Virtual Network Computing"
.SH NAME
vncconfig \- configure and control a VNC server
.SH SYNOPSIS
.B vncconfig
.RI [ parameters ]
.br
.B vncconfig
.RI [ parameters ]
.B \-connect \fP[\fB-view-only\fP]
.IR host [: port ]
.br
.B vncconfig
.RI [ parameters ]
.B \-disconnect
.br
.B vncconfig
.RI [ parameters ]
.RB [ -set ]
.IR Xvnc-param = value " ..."
.br
.B vncconfig
.RI [ parameters ]
.B \-list
.br
.B vncconfig
.RI [ parameters ]
\fB\-get\fP \fIXvnc-param\fP
.br
.B vncconfig
.RI [ parameters ]
\fB\-desc\fP \fIXvnc-param\fP
.SH DESCRIPTION
.B vncconfig
is used to configure and control a running instance of Xvnc, or any other X
server with the VNC extension.  Note that it cannot be used to control VNC
servers prior to version 4.

When run with no options, it runs as a kind of "helper" application for Xvnc.
Its main purpose when run in this mode is to query the user how new
connections should be handled (provided this feature is enabled). The
\fB-nowin\fP flag can be used if you always want the query support but don't
wish to clutter the desktop with the settings window - alternatively the
\fB-iconic\fP option can be used to make it iconified by default.

When run in any other mode, \fBvncconfig\fP is a one-shot program used to
configure or control Xvnc as appropriate.  It can be used to tell Xvnc to
connect or disconnect from listening viewers, and to set and retrieve Xvnc's
parameters.

Note that the DISPLAY environment variable or the \fB\-display\fP option
must be set as appropriate to control Xvnc.  If you run it on an ordinary X
server (or on a version 3 Xvnc) you will get an error message saying that there
is no VNC extension.

.SH OPTIONS
.TP
.B \-connect \fP[\fB-view-only\fP] \fIhost\fP[:\fIport\fP]
Tells an Xvnc server to make a "reverse" connection to a listening VNC viewer
(normally connections are made the other way round - the viewer connects to the
server). \fIhost\fP is the host where the listening viewer is running. If it's
not listening on the default port of 5500, you can specify \fIhost:port\fP
instead. The \fB-view-only\fP option specifies that the server must ignore all
keyboard or mouse events sent by the client.
.
.TP
.B \-desc \fIXvnc-param\fP
Prints a short description of the given Xvnc parameter.
.
.TP
.B \-disconnect
This causes Xvnc to disconnect from all viewers so that the VNC desktop is not
displayed anywhere.
.
.TP
.B \-get \fIXvnc-param\fP
Prints the current value of the given Xvnc parameter.
.
.TP
.B \-list
Lists all the parameters supported by Xvnc.
.
.TP
[\fB-set\fP] \fIXvnc-param\fP=\fIvalue\fP
Sets an Xvnc parameter to the given value.  Note that some of Xvnc's parameters
are read only once at startup so that changing them in this way may not have
any effect.

.SH PARAMETERS
.B vncconfig
also has parameters of its own which can be set on the command line.  These
should not be confused with Xvnc's parameters which are manipulated with the
\fB-set\fP, \fB-get\fP, \fB-list\fP and \fB-desc\fP options.

Parameters can be turned on with -\fIparam\fP or off with -\fIparam\fP=0.
Parameters which take a value can be specified as -\fIparam\fP \fIvalue\fP.
Other valid forms are \fIparam\fP\fB=\fP\fIvalue\fP -\fIparam\fP=\fIvalue\fP
--\fIparam\fP=\fIvalue\fP.  Parameter names are case-insensitive.

.TP
.B \-display \fIXdisplay\fP
Specifies the Xvnc server to control.
.
.TP
.B \-iconic
When run as a "helper" app, make the window iconified at startup.
.
.TP
.B \-nowin
When run as a "helper" app, don't put up a window.

.SH SEE ALSO
.BR vncpasswd (1),
.BR vncviewer (1),
.BR Xvnc (1),
.BR vncsession (8)
.br
https://www.tigervnc.org

.SH AUTHOR
Tristan Richardson, RealVNC Ltd. and others.

VNC was originally developed by the RealVNC team while at Olivetti
Research Ltd / AT&T Laboratories Cambridge.  TightVNC additions were
implemented by Constantin Kaplinsky. Many other people have since
participated in development, testing and support. This manual is part
of the TigerVNC software suite.


unix/vncconfig/vncconfig.cxx
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright 2012-2016 Pierre Ossman for Cendio AB
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */
//
// VNC server configuration utility
//

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>

#include "vncExt.h"
#include <X11/X.h>
#include <X11/Xatom.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>
#include <signal.h>

#include <core/Configuration.h>
#include <core/Exception.h>
#include <core/LogWriter.h>
#include <core/Logger_stdio.h>

#include "QueryConnectDialog.h"
#include "TXCheckbox.h"
#include "TXLabel.h"
#include "TXWindow.h"

static core::LogWriter vlog("vncconfig");

core::StringParameter displayname("display", "The X display", "");
core::BoolParameter noWindow("nowin", "Don't display a window", 0);
core::BoolParameter iconic("iconic", "Start with window iconified", 0);

#define ACCEPT_CUT_TEXT "AcceptCutText"
#define SEND_CUT_TEXT "SendCutText"

#define SET_PRIMARY "SetPrimary"
#define SEND_PRIMARY "SendPrimary"

char* programName = nullptr;
Display* dpy;
int vncExtEventBase, vncExtErrorBase;

static bool getBoolParam(Display* dpy_, const char* param) {
  char* data;
  int len;
  if (XVncExtGetParam(dpy_, param, &data, &len)) {
    if (strcmp(data, "1") == 0)
      return true;
  }
  return false;
}

class VncConfigWindow : public TXWindow,
                        public TXEventHandler,
                        public TXDeleteWindowCallback,
                        public TXCheckboxCallback,
                        public QueryResultCallback {
public:
  VncConfigWindow(Display* dpy_)
      : TXWindow(dpy_, 300, 100), acceptClipboard(dpy_, "Accept clipboard from viewers", this, false, this),
        setPrimaryCB(dpy_, "Also set primary selection", this, false, this),
        sendClipboard(dpy_, "Send clipboard to viewers", this, false, this),
        sendPrimaryCB(dpy_, "Send primary selection to viewers", this, false, this), queryConnectDialog(nullptr) {
    int y = yPad;
    acceptClipboard.move(xPad, y);
    acceptClipboard.checked(getBoolParam(dpy, ACCEPT_CUT_TEXT));
    y += acceptClipboard.height();
    setPrimaryCB.move(xPad + 10, y);
    setPrimaryCB.checked(getBoolParam(dpy, SET_PRIMARY));
    setPrimaryCB.disabled(!acceptClipboard.checked());
    y += setPrimaryCB.height();
    sendClipboard.move(xPad, y);
    sendClipboard.checked(getBoolParam(dpy, SEND_CUT_TEXT));
    y += sendClipboard.height();
    sendPrimaryCB.move(xPad + 10, y);
    sendPrimaryCB.checked(getBoolParam(dpy, SEND_PRIMARY));
    sendPrimaryCB.disabled(!sendClipboard.checked());
    y += sendPrimaryCB.height();
    setEventHandler(this);
    toplevel("VNC config", this, 0, nullptr, nullptr, iconic);
    XVncExtSelectInput(dpy, win(), VncExtQueryConnectMask);
  }

  // handleEvent()

  void handleEvent(TXWindow* /*w*/, XEvent* ev) override {
    if (ev->type == vncExtEventBase + VncExtQueryConnectNotify) {
      vlog.debug("Query connection event");
      if (queryConnectDialog)
        delete queryConnectDialog;
      queryConnectDialog = nullptr;
      char* qcAddress;
      char* qcUser;
      int qcTimeout;
      if (XVncExtGetQueryConnect(dpy, &qcAddress, &qcUser, &qcTimeout, &queryConnectId)) {
        if (qcTimeout)
          queryConnectDialog = new QueryConnectDialog(dpy, qcAddress, qcUser, qcTimeout, this);
        if (queryConnectDialog)
          queryConnectDialog->map();
        XFree(qcAddress);
        XFree(qcUser);
      }
    }
  }

  // TXDeleteWindowCallback method
  void deleteWindow(TXWindow* /*w*/) override {
    exit(1);
  }

  // TXCheckboxCallback method
  void checkboxSelect(TXCheckbox* checkbox) override {
    if (checkbox == &acceptClipboard) {
      XVncExtSetParam(dpy, ACCEPT_CUT_TEXT, acceptClipboard.checked() ? "1" : "0");
      setPrimaryCB.disabled(!acceptClipboard.checked());
    } else if (checkbox == &sendClipboard) {
      XVncExtSetParam(dpy, SEND_CUT_TEXT, sendClipboard.checked() ? "1" : "0");
      sendPrimaryCB.disabled(!sendClipboard.checked());
    } else if (checkbox == &setPrimaryCB) {
      XVncExtSetParam(dpy, SET_PRIMARY, setPrimaryCB.checked() ? "1" : "0");
    } else if (checkbox == &sendPrimaryCB) {
      XVncExtSetParam(dpy, SEND_PRIMARY, sendPrimaryCB.checked() ? "1" : "0");
    }
  }

  // QueryResultCallback interface
  void queryApproved() override {
    XVncExtApproveConnect(dpy, queryConnectId, 1);
  }
  void queryRejected() override {
    XVncExtApproveConnect(dpy, queryConnectId, 0);
  }

private:
  TXCheckbox acceptClipboard, setPrimaryCB;
  TXCheckbox sendClipboard, sendPrimaryCB;

  QueryConnectDialog* queryConnectDialog;
  void* queryConnectId;
};

static void usage() {
  fprintf(stderr, "Usage: %s [parameters]\n", programName);
  fprintf(stderr,
          "       %s [parameters] -connect "
          "[-view-only] <host>[:<port>]\n",
          programName);
  fprintf(stderr, "       %s [parameters] -disconnect\n", programName);
  fprintf(stderr, "       %s [parameters] [-set] <Xvnc-param>=<value> ...\n", programName);
  fprintf(stderr, "       %s [parameters] -list\n", programName);
  fprintf(stderr, "       %s [parameters] -get <param>\n", programName);
  fprintf(stderr, "       %s [parameters] -desc <param>\n", programName);
  fprintf(stderr, "\n"
                  "Parameters can be turned on with -<param> or off with -<param>=0\n"
                  "Parameters which take a value can be specified as "
                  "-<param> <value>\n"
                  "Other valid forms are <param>=<value> -<param>=<value> "
                  "--<param>=<value>\n"
                  "Parameter names are case-insensitive.  The parameters are:\n\n");
  core::Configuration::listParams(79, 14);
  exit(1);
}

void removeArgs(int* argc, char** argv, int first, int n) {
  if (first + n > *argc)
    return;
  for (int i = first + n; i < *argc; i++)
    argv[i - n] = argv[i];
  *argc -= n;
}

int main(int argc, char** argv) {
  programName = argv[0];
  core::initStdIOLoggers();
  core::LogWriter::setLogParams("*:stderr:30");

  // Process vncconfig's own parameters first, then we process the
  // other arguments when we have the X display.
  int i;
  for (i = 1; i < argc;) {
    int ret;

    ret = core::Configuration::handleParamArg(argc, argv, i);
    if (ret > 0) {
      i += ret;
      continue;
    }

    if (strcmp(argv[i], "-help") == 0) {
      usage();
    }

    if (strcmp(argv[i], "-version") == 0) {
      fprintf(stderr, "vncconfig (TigerVNC) %s\n", PACKAGE_VERSION);
      exit(0);
    }

    break;
  }

  if (!(dpy = XOpenDisplay(displayname))) {
    fprintf(stderr, "%s: unable to open display \"%s\"\n", programName, XDisplayName(displayname));
    exit(1);
  }

  if (!XVncExtQueryExtension(dpy, &vncExtEventBase, &vncExtErrorBase)) {
    fprintf(stderr, "No VNC extension on display %s\n", XDisplayName(displayname));
    exit(1);
  }

  if (i < argc) {
    for (; i < argc; i++) {
      if (strcmp(argv[i], "-connect") == 0) {
        Bool viewOnly = False;
        i++;
        if (strcmp(argv[i], "-view-only") == 0) {
          viewOnly = True;
          i++;
        }
        if (i >= argc)
          usage();
        if (!XVncExtConnect(dpy, argv[i], viewOnly)) {
          fprintf(stderr, "Connecting to %s failed\n", argv[i]);
        }
      } else if (strcmp(argv[i], "-disconnect") == 0) {
        if (!XVncExtConnect(dpy, "", False)) {
          fprintf(stderr, "Disconnecting all clients failed\n");
        }
      } else if (strcmp(argv[i], "-get") == 0) {
        i++;
        if (i >= argc)
          usage();
        char* data;
        int len;
        if (XVncExtGetParam(dpy, argv[i], &data, &len)) {
          printf("%.*s\n", len, data);
        } else {
          fprintf(stderr, "Getting param %s failed\n", argv[i]);
        }
        XFree(data);
      } else if (strcmp(argv[i], "-desc") == 0) {
        i++;
        if (i >= argc)
          usage();
        char* desc = XVncExtGetParamDesc(dpy, argv[i]);
        if (desc) {
          printf("%s\n", desc);
        } else {
          fprintf(stderr, "Getting description for param %s failed\n", argv[i]);
        }
        XFree(desc);
      } else if (strcmp(argv[i], "-list") == 0) {
        int nParams;
        char** list = XVncExtListParams(dpy, &nParams);
        for (int n = 0; n < nParams; n++) {
          printf("%s\n", list[n]);
        }
        XVncExtFreeParamList(list);
      } else if (strcmp(argv[i], "-set") == 0) {
        i++;
        if (i >= argc)
          usage();

        char* equal = strchr(argv[i], '=');
        if (!equal) {
          fprintf(stderr, "%s: Invalid parameter syntax '%s'\n", programName, argv[i]);
          fprintf(stderr, "See '%s -help' for more information.\n", programName);
          exit(1);
        }

        std::string name(argv[i], equal - argv[i]);
        std::string value(equal + 1);

        if (!XVncExtSetParam(dpy, name.c_str(), value.c_str()))
          fprintf(stderr, "Setting param %s failed\n", argv[i]);
      } else if (argv[i][0] == '-') {
        fprintf(stderr, "%s: Unrecognized option '%s'\n", programName, argv[i]);
        fprintf(stderr, "See '%s -help' for more information.\n", programName);
        exit(1);
      } else {
        char* equal = strchr(argv[i], '=');
        if (!equal) {
          fprintf(stderr, "%s: Invalid parameter syntax '%s'\n", programName, argv[i]);
          fprintf(stderr, "See '%s -help' for more information.\n", programName);
          exit(1);
        }

        std::string name(argv[i], equal - argv[i]);
        std::string value(equal + 1);

        if (!XVncExtSetParam(dpy, name.c_str(), value.c_str()))
          fprintf(stderr, "Setting param %s failed\n", argv[i]);
      }
    }

    return 0;
  }

  try {
    TXWindow::init(dpy, "Vncconfig");

    VncConfigWindow w(dpy);
    if (!noWindow)
      w.map();

    while (true) {
      struct timeval tv;
      struct timeval* tvp = nullptr;

      // Process any incoming X events
      TXWindow::handleXEvents(dpy);

      // Process expired timers and get the time until the next one
      int timeoutMs = core::Timer::checkTimeouts();
      if (timeoutMs >= 0) {
        tv.tv_sec = timeoutMs / 1000;
        tv.tv_usec = (timeoutMs % 1000) * 1000;
        tvp = &tv;
      }

      // If there are X requests pending then poll, don't wait!
      if (XPending(dpy)) {
        tv.tv_usec = tv.tv_sec = 0;
        tvp = &tv;
      }

      // Wait for X events, VNC traffic, or the next timer expiry
      fd_set rfds;
      FD_ZERO(&rfds);
      FD_SET(ConnectionNumber(dpy), &rfds);
      int n = select(FD_SETSIZE, &rfds, nullptr, nullptr, tvp);
      if (n < 0)
        throw core::socket_error("select", errno);
    }

    XCloseDisplay(dpy);

  } catch (std::exception& e) {
    vlog.error("%s", e.what());
  }

  return 0;
}
